n-dimensional rescale, resize, and pyramid transforms

grlee77 opened the pull request:
This is a pretty simple pull request extending n-dimensional support to the following transforms:
- `skimage.transform.rescale`
- `skimage.transform.resize`
- `skimage.transform.pyramid_*`

This is useful for me as I work primarily in medical imaging where data is often either 3D volumetric or 3D+time.

As was previously the case, resize still calls the two-dimensional interpolations implemented in `_shared/interpolation.pyx` for 2D or 2D+color images, while all other cases use `scipy.ndimage.map_coordinates`.  Previously `map_coordinates` was set up explicitly for 3D only.

The changes to `resize` are substantially smaller than the diff would indicate at first glance, because I switched the order of the  if / else statement calling the 2D or n-D interpolation code.  (The logic for when to call 2D seemed slightly simpler to express)

A multichannel argument was added to `rescale` to indicate that the last axis should not be reshaped.  This defaults to True for 3D inputs to replicate the previous behaviour for 2D+color images.  To do a true 3D rescale, the user must explicitly specify `multichannel=False`.

Similarly in `pyramids.py`:
The `_smooth` function previously always looped 2D smoothing operations over the last axis if the input was 3D.  Now there is a new `multichannel` argument that explicitly controls this behavior of looping over the last axis, regardless of image dimension.  As for `rescale` above, `multichannel` defaults to None which will result in a call to a small function to replicate the old behaviour of True for 3D inputs and False otherwise.

ahojnnes commented on skimage/transform/pyramids.py:
```diff
@@ -50,6 +52,8 @@ def pyramid_reduce(image, downscale=2, sigma=None, order=1,
         cval is the value when mode is equal to 'constant'.
     cval : float, optional
         Value to fill past edges of input if mode is 'constant'.
+        If True, last axis will not be rescaled.  The default is True for 3D
```
sth missing here

jni commented on skimage/transform/_warps.py:
```diff
@@ -6,6 +6,14 @@
+    if ndim == 3:
+        return True
+    else:
+        return False
```
We've discussed in the past that this is not a magic preferred by the team... But in this case it's different because you're not adding any magic, it was already there. However, the current selector is really imperfect because it will fail quite dramatically for 3D grayscale images. (Note also that you are ignoring the `multichannel` argument!!!)

I see two options for this:

1. If we want to maintain this magic, which to reiterate was already there:

```python
def _multichannel_default(multichannel, ndim, last_dim):
    if multichannel is not None:
        return multichannel
    if ndim == 3 and last_dim in [3, 4]:
        return True
    else:
        return False
```

2. if we want to deprecate this behaviour and specify a new default (i.e. universally `True` or `False`), we additionally need to raise a deprecation warning when `multichannel is None`.

stefanv replied:
:+1: for (deprecate and make consistent)

grlee77 replied:
@jni:  In an earlier commit I had it nearly identical to what you propose above.  For some reason I later changed it so that `_multichannel_default` is only ever called when `multichannel` is None.  

I can put it back as you propose for clarity and add the deprecation warning

grlee77 replied:
@stefanv:  One other related question: If I add a deprication warning as above, then several existing tests fail due to the warning being raised.  Should I modify these tests to explicitly set the `multichannel` argument?  (or is there a way to tell the tests to ignore that warning so we can make sure the default ``multichannel=None`` result matches prior versions?)

stefanv replied:
The tests are set up to fail on warnings so that those don't slip by unnoticed, but if you look at some of the other tests, you'll see that we have utility functions (``skimage._shared._warnings``) that can take care of it.

jni commented on skimage/transform/_warps.py:
```diff
@@ -58,36 +66,25 @@ def resize(image, output_shape, order=1, mode='constant', cval=0, clip=True,
+        orig_shape = orig_shape[:-1]
+    orig_shape = np.asarray(orig_shape)
+
+    dim_scales = orig_shape.astype('f8') / output_shape
```
Urm, why not `orig_shape = np.array(orig_shape, dtype=float)` above, to avoid this `'f8'` business?

Indeed I have no idea what you're doing in this whole paragraph; you might want to sprinkle a few comments in there. =)

grlee77 replied:
I agree that section is confusing and will try to clarify/simplify it.  

the 'f8' that was a lazy shorthand for np.float64.  Making sure one of the arrays is floating avoids an integer division problem there for python 2.x.

grlee77 commented:
Thank you all for the quick reviews on this pull request as well as the previous one!

grlee77 commented:
In addition to the changes suggested above, I explicitly set the multichannel argument in a few other locations in the codebase where `rescale` or pyramids were being used so that the `multichannel` deprecation warning will not be raised.  There were only a handful of these.

grlee77 commented:
All seems well with the tests.  Let me know if there is anything else to be done here.

It looks like I may need update the ndimage conventions in accord with #1550 and possibly change the default mode to 'reflect' as discussed in #1514.

grlee77 commented:
It has been awhile, but I would like to revisit this PR if there is still interest

a made a small, separate PR for the default edge mode changes in #2484 

It would probably make sense to rebase this PR on top of that one

pep8speaks commented:
Hello @grlee77! Thanks for updating the PR.



 - In the file [`skimage/feature/orb.py`](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/feature/orb.py), following are the PEP8 issues :

> [Line 8:30](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/feature/orb.py#L8): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 323:80](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/feature/orb.py#L323): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 324:80](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/feature/orb.py#L324): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)

 - In the file [`skimage/transform/_warps.py`](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/transform/_warps.py), following are the PEP8 issues :

> [Line 11:80](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/transform/_warps.py#L11): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (87 > 79 characters)
> [Line 39:80](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/transform/_warps.py#L39): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (80 > 79 characters)
> [Line 589:25](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/transform/_warps.py#L589): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 608:80](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/transform/_warps.py#L608): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (86 > 79 characters)
> [Line 774:55](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/transform/_warps.py#L774): [E502](https://duckduckgo.com/?q=pep8%20E502) the backslash is redundant between brackets
> [Line 775:19](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/transform/_warps.py#L775): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 783:34](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/transform/_warps.py#L783): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 784:34](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/transform/_warps.py#L784): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent

 - In the file [`skimage/transform/pyramids.py`](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/transform/pyramids.py), following are the PEP8 issues :

> [Line 268:80](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/transform/pyramids.py#L268): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (101 > 79 characters)



 - In the file [`skimage/transform/tests/test_radon_transform.py`](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/transform/tests/test_radon_transform.py), following are the PEP8 issues :

> [Line 157:80](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/transform/tests/test_radon_transform.py#L157): [E501](https://duckduckgo.com/?q=pep8%20E501) line too long (91 > 79 characters)
> [Line 247:5](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/transform/tests/test_radon_transform.py#L247): [E731](https://duckduckgo.com/?q=pep8%20E731) do not assign a lambda expression, use a def

 - In the file [`skimage/transform/tests/test_warps.py`](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/transform/tests/test_warps.py), following are the PEP8 issues :

> [Line 50:5](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/transform/tests/test_warps.py#L50): [E731](https://duckduckgo.com/?q=pep8%20E731) do not assign a lambda expression, use a def
> [Line 323:37](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/transform/tests/test_warps.py#L323): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 323:60](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/transform/tests/test_warps.py#L323): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 358:28](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/transform/tests/test_warps.py#L358): [E201](https://duckduckgo.com/?q=pep8%20E201) whitespace after '['
> [Line 364:31](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/transform/tests/test_warps.py#L364): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ','
> [Line 365:28](https://github.com/grlee77/scikit-image/blob/3d5717a01bccb6d5ab14e269eeea0b86e7bb3eda/skimage/transform/tests/test_warps.py#L365): [E201](https://duckduckgo.com/?q=pep8%20E201) whitespace after '['

grlee77 commented:
I rebased this after the pytest merge.  I dropped a single commit related to older numpy compatibility (cbf62401a696f735956912c5538b3b2ca318f2ec) as it is no longer necessary with the new numpy 1.11 minimum requirement.

The final commit is a bunch of random PEP8 fixes of issues brought up by pep8speaks.

The individual commit history is definitely not ideal, as I was still new to learning git workflows at the time.  If it is desired to go forward with this PR I can rebase into a handful of more logical commits that may make it a bit easier to review?

jni commented on skimage/transform/_warps.py:
```diff
@@ -18,14 +17,27 @@
+        return multichannel
+    else:
+        warn('The default multichannel argument (None) is deprecated. '
+             'Specifiy True or False explicitly.')
```
I suspect we want multichannel=False in the future but the current behavior is True for 3D images and False for 2D or 4D+ images? I think somehow this has to come across in the deprecation message. Otherwise multichannel would be a required argument after the deprecation, which doesn't sound good. =)

Oh and btw, specify in which version the deprecation will take place.

grlee77 replied:
The current behavior in `0.13` treats as multichannel when the input is 3D **and** `output_shape[2] == image.shape[2]`.  Otherwise it is not treated as multichannel.  I think `multichannel_default` should be updated to match that behavior backwards compatibility.

I will update the deprecation message appropriately assuming the future default will be `False`.  Assuming this goes into release 0.14, deprecation would be 0.16?

grlee77 replied:
actually `_multichannel_default` is fine as is (aside from the warnings text).  It is only used by `rescale` and the various pyramid functions which all previously assumed 2D+channels if the input was 3D and would error on inputs > 3D.

jni commented on skimage/transform/_warps.py:
```diff
@@ -84,48 +96,40 @@ def resize(image, output_shape, order=1, mode=None, cval=0, clip=True,
+    ndim_out = len(output_shape)
+    input_shape = image.shape
+    if ndim_out > image.ndim:
+        # append dimensions to input_shape
```
How did you decide to append instead of prepend? This is the opposite of numpy broadcasting rules, so I would want a justification noted in a comment here...

grlee77 replied:
I pref disallowing output_shape with a different number of dimensions than the input.  The problem with this is that in `0.13` it is already allowed to give a 2D input with 3D output shape.  The additional dimension gets appended to the end:
https://github.com/scikit-image/scikit-image/blob/d020c22780bdce25546f55be293bb5eeae92ac0e/skimage/transform/_warps.py#L98-L99

jni commented on skimage/transform/_warps.py:
```diff
@@ -84,48 +96,40 @@ def resize(image, output_shape, order=1, mode=None, cval=0, clip=True,
-
-    else:  # 2-dimensional interpolation
-
+    output_shape = np.asarray(output_shape)
```
Why this conversion? It complicates downstream syntax, forcing a call to `np.concatenate`. I suggest making it a tuple.

grlee77 replied:
I can change to `output_shape = tuple(output_shape)`.  That should also convert a 1D array to a tuple and allow the simpler tuple concatenation.

jni commented on skimage/transform/_warps.py:
```diff
@@ -84,48 +96,40 @@ def resize(image, output_shape, order=1, mode=None, cval=0, clip=True,
+        image = np.reshape(image, input_shape)
+    elif ndim_out == image.ndim - 1:
+        # multichannel case: append shape of last axis
+        output_shape = np.concatenate((output_shape, [image.shape[-1], ]))
```
If `output_shape` is a tuple, this becomes `output_shape += (image.shape[-1],)`

jni commented on skimage/transform/_warps.py:
```diff
@@ -134,11 +138,30 @@ def resize(image, output_shape, order=1, mode=None, cval=0, clip=True,
 
+    else:  # n-dimensional interpolation
+        for i in range(ndim_out - image.ndim):
+            image = image[..., np.newaxis]
```
Didn't we already reshape the image above? And if so, is this not redundant?

grlee77 replied:
True.  I moved it above and then never removed it from here :)

jni commented on skimage/transform/_warps.py:
```diff
@@ -178,6 +201,9 @@ def rescale(image, scale, order=1, mode=None, cval=0, clip=True,
         image is converted according to the conventions of `img_as_float`.
+    multichannel : bool, optional
+        If True, last axis will not be rescaled.  The default is True for 3D
+        (2D+color) inputs, False otherwise.
```
... but will be False everywhere in version 0.16? I think overall we've decided to get rid of all multichannel guessing throughout the library on the way to 1.0.

soupault replied:
:+1: I agree: we should add deprecation message if `multichannel is None` saying that it will default to `False` in 0.16.

soupault replied:
I'd rephrase it like this: ` By default, is set to True for 3D (2D+color) inputs, and False for others.`.

jni commented on skimage/transform/_warps.py:
```diff
@@ -190,16 +216,14 @@ def rescale(image, scale, order=1, mode=None, cval=0, clip=True,
+    multichannel = _multichannel_default(multichannel, image.ndim)
+    scale = np.atleast_1d(scale)
+    if len(scale) > 1 and len(scale) != image.ndim:
+        raise ValueError("must supply a single scale or one value per axis.")
```
Wouldn't there be a third case — for multichannel images, len(scale) == image.ndim[:-1], right?

jni commented on skimage/transform/pyramids.py:
```diff
@@ -3,16 +3,17 @@
-        for dim in range(image.shape[2]):
+    # apply Gaussian filter to all channels independently
+    if multichannel:
+        for dim in range(image.shape[-1]):
```
This seems unnecessarily complex — isn't it equivalent to run a single gaussian filter with the corresponding sigma = 0?

jni commented on skimage/transform/pyramids.py:
```diff
@@ -273,31 +279,32 @@ def pyramid_laplacian(image, max_layer=-1, downscale=2, sigma=None, order=1,
 
     # build downsampled images until max_layer is reached or downscale process
     # does not change image size
     while layer != max_layer:
```
I know it's not on you but this should clearly be a for loop. =)

jni commented on skimage/transform/tests/test_warps.py:
```diff
@@ -47,7 +47,8 @@ def test_warp_callable():
 
-    shift = lambda xy: xy + 1
+    def shift(xy):
+        return xy + 1
```
Nice. ;)

jni commented:
Ah, here is the discussion, which was sort of cut off by @noahstier but I think is worth hashing out in a longer discussion. =)

https://github.com/scikit-image/scikit-image/pull/2587#issuecomment-291081152

soupault commented on skimage/transform/_warps.py:
```diff
@@ -18,14 +17,27 @@
+
+
 def resize(image, output_shape, order=1, mode=None, cval=0, clip=True,
            preserve_range=False):
```
We should add `multichannel` also to `resize`.

grlee77 replied:
What would it be used for?  I think the two shapes specify everything that needs to be known?  I guess if multichannel was specified an image.shape[-1] != output_shape[-1] and error could be raised.

soupault replied:
Well, both `resize` and `rescale` are exposed to the user. They are used for the same purpose, therefore, I think, they should have the similar API. There is nothing more than that :).

In my opinion, the thinking should be the following: if a function performs any logic on "multichanneling", the argument has to be explicitly provided.

jni replied:
@soupault but the API is necessarily different — rescale can take float input, while resize *always* needs a tuple. This means that the multichannel nature can be inferred from the output and input shapes. At any rate, I consider this something that can be fixed post merge if necessary.

soupault replied:
> This means that the multichannel nature can be inferred from the output and input shapes.
A-a-a-and we are back to it :).

@jni I understand your point. My feeling is that the API should be as homogeneous as possible.
Maybe we could just have another input from @scikit-image/core ?

jni replied:
@soupault yes I'm happy to discuss it. Notice that I consider this a different discussion to what we've had before: when the argument can be inferred unambiguously, vs when we are just guessing.

At any rate, I'm happy to have the discussion, but I don't think it should hold this PR up. Perhaps you can raise an issue afterwards? And it can cover existing cases of guessing.

grlee77 replied:
I have not added it so far as I am not sure what the logic would be within the function.  As @jni stated, whether or not the last axis would be reshaped is unambiguous from `image.shape` and `output_shape`.  I suppose if `multichannel=True` we could raise an error if `image.shape[-1] != output_shape[-1]`, but I'm not sure what else it could do?

soupault replied:
Alright. As far as both of you are strongly concerned not to change the API for `resize`, let's move the discussion to a separate thread and try to work out the best solution there.

soupault commented on skimage/transform/pyramids.py:
```diff
@@ -29,7 +30,7 @@ def _check_factor(factor):
 
 def pyramid_reduce(image, downscale=2, sigma=None, order=1,
-                   mode='reflect', cval=0):
+                   mode='reflect', cval=0, multichannel=None):
```
Please, add deprecation message here as well.

grlee77 replied:
For all of these pyramid functions, the message is already there within the call to the `_multichannel_default` routine.

soupault commented on skimage/transform/pyramids.py:
```diff
@@ -61,29 +65,27 @@ def pyramid_reduce(image, downscale=2, sigma=None, order=1,
 
 def pyramid_expand(image, upscale=2, sigma=None, order=1,
-                   mode='reflect', cval=0):
+                   mode='reflect', cval=0, multichannel=None):
```
And here.

soupault commented on skimage/transform/pyramids.py:
```diff
@@ -199,23 +204,21 @@ def pyramid_gaussian(image, max_layer=-1, downscale=2, sigma=None, order=1,
 
 def pyramid_laplacian(image, max_layer=-1, downscale=2, sigma=None, order=1,
-                      mode='reflect', cval=0):
+                      mode='reflect', cval=0, multichannel=None):
```
And here.

grlee77 commented:
Most of the comments have been addressed.  I think the ones still outstanding are:

- ongoing discussion on whether to use `multichannel` argument in all of these cases

- whether to add a new parameter to `resize` that lets to the user to choose whether to append or preprend dimensions when the output_shape has more dimensions than the input?  Another alternative is to deprecate allowing `output_shape` to be a different length than the number of image dimensions.  I think I am slightly in favor of the latter.

- does `resize` also need a multichannel argument?  I did not see a need for it, but adding it was suggested above.

soupault commented on skimage/transform/_warps.py:
```diff
@@ -18,24 +17,38 @@
 )
 
 
+def _multichannel_default(multichannel, ndim):
```
Please, add a reminder what should be changed/removed by 0.16 to https://github.com/scikit-image/scikit-image/blob/master/TODO.txt.

soupault commented on skimage/transform/_warps.py:
```diff
@@ -84,48 +97,40 @@ def resize(image, output_shape, order=1, mode=None, cval=0, clip=True,
+    output_shape = tuple(output_shape)
+    ndim_out = len(output_shape)
+    input_shape = image.shape
+    if ndim_out > image.ndim:
```
@grlee77 I assume, we both agree to get rid of this?
Does anyone object?

jni replied:
@soupault no, this is necessary for backwards compatibility — see above

jni replied:
(above meaning the code that has been removed)

soupault replied:
Ah, yes, I've missed this point.

soupault commented on skimage/transform/_warps.py:
```diff
@@ -178,6 +199,9 @@ def rescale(image, scale, order=1, mode=None, cval=0, clip=True,
         Whether to keep the original range of values. Otherwise, the input
         image is converted according to the conventions of `img_as_float`.
+    multichannel : bool, optional
+        By default, is set to True for 3D (2D+color) inputs, and False for
```
Sorry, I meant to add this reformulated sentence in addition to the description. It's not clear what this argument is for as it is.

soupault commented on skimage/transform/_warps.py:
```diff
@@ -134,11 +139,27 @@ def resize(image, output_shape, order=1, mode=None, cval=0, clip=True,
     return out
 
 
 def rescale(image, scale, order=1, mode=None, cval=0, clip=True,
```
Please, update this docstring as well:
`(row_scale, col_scale)` -> nD notation.

sciunto commented on skimage/transform/pyramids.py:
```diff
@@ -47,8 +47,10 @@ def pyramid_reduce(image, downscale=2, sigma=None, order=1,
     multichannel : bool, optional
-        By default, is set to True for 3D (2D+color) inputs, and False for
-        others.  Starting in release 0.16, this will always default to False.
+        hether the last axis of the image is to be interpreted as multiple
```
Whether

sciunto commented on skimage/transform/pyramids.py:
```diff
@@ -47,8 +47,10 @@ def pyramid_reduce(image, downscale=2, sigma=None, order=1,
-        By default, is set to True for 3D (2D+color) inputs, and False for
-        others.  Starting in release 0.16, this will always default to False.
+        hether the last axis of the image is to be interpreted as multiple
+        channels or another spatial dimension.  By default, is set to True for
```
two spaces instead of one after .

sciunto commented on skimage/transform/pyramids.py:
```diff
@@ -47,8 +47,10 @@ def pyramid_reduce(image, downscale=2, sigma=None, order=1,
-        others.  Starting in release 0.16, this will always default to False.
+        hether the last axis of the image is to be interpreted as multiple
+        channels or another spatial dimension.  By default, is set to True for
+        3D (2D+color) inputs, and False for others.  Starting in release 0.16,
```
two spaces instead of one after .

sciunto commented on skimage/transform/_warps.py:
```diff
@@ -200,8 +200,10 @@ def rescale(image, scale, order=1, mode=None, cval=0, clip=True,
-        others.  Starting in release 0.16, this will always default to False.
+        Whether the last axis of the image is to be interpreted as multiple
+        channels or another spatial dimension.  By default, is set to True for
+        3D (2D+color) inputs, and False for others.  Starting in release 0.16,
```
two spaces instead of one after .

grlee77 replied:
okay.  fixing these now

sciunto commented on skimage/transform/_warps.py:
```diff
@@ -200,8 +200,10 @@ def rescale(image, scale, order=1, mode=None, cval=0, clip=True,
-        By default, is set to True for 3D (2D+color) inputs, and False for
-        others.  Starting in release 0.16, this will always default to False.
+        Whether the last axis of the image is to be interpreted as multiple
+        channels or another spatial dimension.  By default, is set to True for
```
two spaces instead of one after .

sciunto commented on skimage/transform/pyramids.py:
```diff
@@ -102,8 +104,11 @@ def pyramid_expand(image, upscale=2, sigma=None, order=1,
     multichannel : bool, optional
-        By default, is set to True for 3D (2D+color) inputs, and False for
-        others.  Starting in release 0.16, this will always default to False.
+        hether the last axis of the image is to be interpreted as multiple
```
Same here :)

sciunto commented on skimage/transform/pyramids.py:
```diff
@@ -169,8 +174,11 @@ def pyramid_gaussian(image, max_layer=-1, downscale=2, sigma=None, order=1,
     multichannel : bool, optional
-        By default, is set to True for 3D (2D+color) inputs, and False for
-        others.  Starting in release 0.16, this will always default to False.
+        hether the last axis of the image is to be interpreted as multiple
```
Again here :P

sciunto commented on skimage/transform/pyramids.py:
```diff
@@ -249,8 +257,11 @@ def pyramid_laplacian(image, max_layer=-1, downscale=2, sigma=None, order=1,
     multichannel : bool, optional
-        By default, is set to True for 3D (2D+color) inputs, and False for
-        others.  Starting in release 0.16, this will always default to False.
+        hether the last axis of the image is to be interpreted as multiple
```
and here...

grlee77 commented:
@soupault 
The remaining stylistic changes have been made and the TODO.txt was updated.  Is there anything else remaining?  I can rebase/squash once everything has been approved.

soupault commented:
Thank you @grlee77 ! As usual, a very high quality PR!
